### ISSUE TITLE
New ewels/multiqc docker image doesn't work with nextflow

### ISSUE THREAD

**Issue by rspreafico** [ewels/MultiQC#935]
**Is your feature request related to a problem? Please describe.**
The official docker container lacks the unix `ps` tool. Therefore, Nextflow cannot run multiqc using docker.

**Describe the solution you'd like**
Add `ps` to the docker container.

**Describe alternatives you've considered**
Na

**Additional context**
Na

**Comment by ewels** [comment]
Hi @rspreafico,

Are you sure about this? When I try running the container I can access `ps`:

```
$ docker run -it --entrypoint /bin/sh ewels/multiqc

/usr/src/multiqc # ps
PID   USER     TIME  COMMAND
    1 root      0:00 /bin/sh
    6 root      0:00 ps
``` 

I tried a minimal nextflow pipeline and got a different error:

```
Command error:
  docker: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "exec: \"/bin/bash\": stat /bin/bash: no such file or directory": unknown.
```

I don't think that `ps` is the culprit here, it's the lack of `bash` (the new MultiQC container is built on alpine that uses `sh` instead of `bash`).

The reason that this has changed recently is because I have refactored the container to be as lightweight as possible - a number of people were complaining that it had a huge filesize. The new version reduced it from around 200MB to ~80MB (which I was feeling quite smug about!). As such, I'm hesitant to return to the previous base image, though that's definitely an option.

In the short term you can use the MultiQC biocontainer image that's built from bioconda (https://biocontainers.pro/#/tools/multiqc), I think that should work.

Phil

**Comment by rspreafico** [comment]
Hi @ewels , thanks for the quick reply.

I was able to reproduce your example and indeed it works. I was experimenting with the container tagged `v1.7` (because for all pipelines I use tagged containers and versioned pipelines, so that they are fully reproducible) and in that case `ps` fails with this command:

```
docker run -it --entrypoint /bin/sh ewels/multiqc:v1.7
# ps
/bin/sh: 1: ps: not found
```

So it does look `ps` is there in the latest, non-tagged MultiQC docker container. Awesome! If we can have the latest version of multiqc (1.7) with a tagged docker container (maybe tagged with something like "v1.7--<build>") that would be great. 

Incidentally, I had tried the biocontainer version (for multiqc 1.7) as well, and it seemed to have the same issue with the latest version of Nextflow (no `ps`) - although I recall distinctly that the multiqc 1.6 biocontainer used to work with an older version of Nextflow. The use of `ps` in Nextflow changed over time, as seen by browsing some Github issues, so there may be a moving target here on both the MultiQC and Nextflow ends.

Bottom line, just need a tagged version of your container with `ps` in it. Thank you! ;-)

**Comment by rspreafico** [comment]
Tried with Nextflow but lack of `bash` is an issue, as you pointed out. I understand the concern about container size, but I recall even early versions of the biocontainers docker images had to add `bash` to the `sh`-only `busybox`, or they couldn't go far. Seems like `bash` is a minimum requisite to do any basic scripting within the container...

**Comment by rspreafico** [comment]
Biocontainers issue: https://github.com/bioconda/bioconda-recipes/issues/13751

**Comment by ewels** [comment]
Hmm, yes this is an issue.

If I switch out the current minimal base for ubuntu (plus installing python), I get these sizes:

```
REPOSITORY                   TAG                 IMAGE ID            CREATED             SIZE
multiqc-slim                 latest              2d9d08b1cdc9        10 minutes ago      201MB
multiqc-ubuntu               latest              334a791afd0c        15 seconds ago      585MB
```

That's quite a jump 😕 Any suggestions on minimal python docker bases that _will_ work with nextflow and the like?

Phil

**Comment by rspreafico** [comment]
(on mobile, please excuse limited insight) is installing only `procps` on the current base image causing such a jump? It is supposed to be a lightweight package with no large dependencies, available both in Ubuntu and Alpine. Just thoughts: installing in the same Docker layer as other packages rather than separately, and not keeping the cache of the package manager should all help with size (probably done already in your Dockerfile).

**Comment by ewels** [comment]
I didn't test adding only `procps` because my container was missing `bash` already.. So I just used `ubuntu` as a base which should have everything. I guess that I should be able to find some kind of mix of Alpine + bash + procps + python which isn't too heavy though. Just need to find that magic combination..! ✨

**Comment by rspreafico** [comment]
Maybe a subset of something like [this](https://github.com/jfloff/alpine-python/blob/master/3.7/Dockerfile)

**Comment by ewels** [comment]
Looks ideal, but then you get lots of installation problems from numpy (required by matplotlib) such as this:
```
  File "numpy/core/setup.py", line 667, in get_mathlib_info
RuntimeError: Broken toolchain: cannot link a simple C program
The command '/bin/sh -c python setup.py install' returned a non-zero code: 1
```

This is why I ended up using a lightweight container specifically for matplotlib before, as someone else had figured out these difficulties already.

**Comment by rspreafico-vir** [comment]
Ok, not sure whether this could help, but I run the `ewels/multiqc:dev` container which is 190mb:

```
docker run --entrypoint /bin/sh --rm -it ewels/multiqc:dev
```

then from within the container:

```
apk add --no-cache bash
```

Then I committed the change with `docker commit` and looked at sizes with `docker images`:

```
test/multiqc                               test                    a82f0ea64953        15 seconds ago      191MB
ewels/multiqc                              dev                     f80503fb7809        3 days ago          190MB
```

`bash` only adds 1mb to the `dev` multiqc container, which already has `ps`. I am sure that making a proper image with `docker build` may add slightly more than the 1 mb added by `docker commit`, but not much more.

So how just adding `bash` to the `dev` container at build & publish time?

**Comment by rspreafico** [comment]
Hi there! Just wondering whether this is still of interest? :-) I can add a PR to add `bash` via the mechanism above if you could point me to the Dockerfile.

**Comment by ewels** [comment]
Yes, definitely! Sorry, I ran out of time and hadn't yet tested it myself yet. The solution above seems almost too good to be true 😆 I had assumed that `bash` would be kind of massive with tonnes of dependencies and much harder to install than just that. But if it works (both standalone and with nextflow), that's brilliant 🎉

**Comment by rspreafico-vir** [comment]
Awesome! Thanks so much!

**Comment by rspreafico** [comment]
BTW forgot to ask, any chance we can get a second build of `v1.7` with this base container? Just because I would like not to put any `dev` tag in a production pipeline ;-) Thanks for considering!

**Comment by ewels** [comment]
Yes - though we may have to do this manually to avoid getting v1.8dev MultiQC code...

**Comment by ewels** [comment]
Hi @rspreafico,

`ewels/multiqc:latest` and `ewels/multiqc:1.7` are both now running the latest container, with MultiQC v1.7 code.

Note that some of the dependencies are newer now, so you get annoying yaml warnings (fixed in v1.8 of MultiQC). But I think it seems to work otherwise.

Would be good to hear how you get on with this! Thanks again for your help.

Phil

**Comment by rspreafico** [comment]
Thanks @ewels !
